### DISCUSSION TITLE
valkey backend?

### DISCUSSION THREAD

**Discussion by dometto** [reclosedev/requests-cache#1038] [General]
Does anyone have experience using this library with [valkey](https://github.com/valkey-io/valkey) as an open source alternative to Redis?

Since valkey is more or less a drop-in replacement for Redis, this should probably work in theory. But since the two projects might diverge more in the future, I'm wondering if there is any interest in a separate Valkey backend.

**Comment by JWCook** [comment]
I'd be interested in maintaining support for valkey, if at some point it diverges in its API or behavior. If those hypothetical future changes just require minor workarounds, I'd prefer to keep it as a single backend class; but if there are major changes that somehow make that infeasible, then we could split that out into a separate backend.

If you or someone else wanted to help out with this, the first step would be to add test containers to both [GitHub Actions config](https://github.com/requests-cache/requests-cache/blob/4eac45fb4a6ccdef563adc1e13ced6df503d9a6e/.github/workflows/build.yml#L49-L56) and [local docker-compose config](https://github.com/requests-cache/requests-cache/blob/main/docker-compose.yml), and then run [Redis integration tests](https://github.com/requests-cache/requests-cache/blob/main/tests/integration/test_redis.py) against both the Redis and valkey containers.

**Reply by dometto** [reply]
I'd be happy to help out with this but I'll have to put it on my backlog and will be able to get to it, realistically, in March.

Just running integration tests against a valkey instance at first makes sense. I think it should also be trivial to adapt the backend to use, configurably, the python redis client, [or the official python valkey client](https://github.com/valkey-io/valkey-py), if that sounds like a good plan.

**Reply by JWCook** [reply]
You can pass a Redis client object to the backend, and I just confirmed that it works interchangeably with a Valkey client:
```py
from valkey import Valkey
from requests_cache import CachedSession, RedisCache

connection = Valkey(host='localhost', port=6379)
backend = RedisCache(connection=connection)
session = CachedSession('http_cache', backend=backend)
r = session.get('https://httpbin.org/get')
r = session.get('https://httpbin.org/get')
print(r)
```

At this point there is probably no need to, unless the user wanted to directly use that internal client object for something valkey-specific that redis-py doesn't support.

**Reply by JWCook** [reply]
I also confirmed that the Redis backend currently works with a Valkey instance, using both `redis-py` (with no code changes required) and `valkey-py` (as in the example above). I added an example to the docs, and added a test container to the docker-compose config.

I manually ran the Redis integration tests against that container, and they all passed, so that looks promising!
